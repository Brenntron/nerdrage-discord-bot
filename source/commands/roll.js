import { min } from "rxjs/operators";

const regex = /!roll\s*(?<count>\d*)(d(?<sides>\d+))?(h(?<hunger>\d+))?/gi

export const rollDie = sides => Math.floor(Math.random() * sides) + 1

export function interpret(dieRolls, { hunger = 0 } = {}) {
  const allDice = [...dieRolls]

  const successes = allDice.filter(n => n >= 6).length
    + (Math.floor(allDice.filter(n => n === 10).length / 2) * 2)

  const isSuccess = successes >= 1
  const criticality = allDice.filter(n => n === 10).length
  const isCritical = Boolean(criticality >= 2)

  const normalDice = [...allDice]
  const hungerDice = [...new Array(hunger)].map(() => normalDice.pop()).filter(Boolean)

  const messiness = !isCritical ? 0 : hungerDice.filter(n => n === 10).length
  const isMessyCritical = Boolean(isCritical && messiness >= 1)

  const bestiality = hungerDice.filter(n => n === 1).length
  const isBestialFailure = Boolean(!isSuccess && bestiality >= 1)

  return {
    allDice,
    bestiality,
    criticality,
    hungerDice,
    isBestialFailure,
    isCritical,
    isMessyCritical,
    isSuccess,
    messiness,
    normalDice,
    successes,
  }
}

const renderNumberArray = array => {
  return array.length
    ? array.sort().reverse().map(n => n === 10 ? '[10]' : n).join(', ')
    : 'none'
}

export function formatMessage(results) {
  let emoji = '👎🏼'
  let resultName = 'Failure'

  if (results.isMessyCritical) {
    emoji = '😬'
    resultName = '[Messy Critical]'
  } else if (results.isCritical) {
    emoji = '👊🏼'
    resultName = 'Critical Success'
  } else if (results.isSuccess) {
    emoji = '👍🏼'
    resultName = 'Success'
  } else if (results.isBestialFailure) {
    emoji = '😈'
    resultName = '[Bestial Failure]'
  }

  return [
    '```css\n',
    `${emoji} ${resultName} (`,
    `${results.successes} total successes`,
    results.isCritical && `, ${results.criticality} crits`,
    results.isBestialFailure > 0 && `, ${results.bestiality} bestiality`,
    `)\n   `,
    `#normal: ${renderNumberArray(results.normalDice)}`,
    results.hungerDice.length > 0 && ` | [hunger]: ${renderNumberArray(results.hungerDice)}`,
    '\n```',
  ].filter(Boolean).join('')
}

export default {
  name: 'roll',
  regex,
  handler(command) {
    const parts = regex.exec(command)
    if (!parts) return

    const settings = {
      count: parseInt(parts.groups.count, 10) || 1,
      sides: parseInt(parts.groups.sides, 10) || 10,
      hunger: parseInt(parts.groups.hunger, 10) || 0,
    }
    const dieRolls = [...new Array(settings.count)].map(() => rollDie(10))

    return formatMessage(interpret(dieRolls, settings))
  }
}
