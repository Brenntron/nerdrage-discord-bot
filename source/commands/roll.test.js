import roll from './roll'

describe('!roll', () => {
  describe('interpret', () => {
    it('computes successes if given a difficulty', () => {
      expect(roll.interpret({ raw: [5, 6, 7], difficulty: 6 }).successes).toBe(2)
      expect(roll.interpret({ raw: [5, 6, 7] }).successes).toBe(undefined)
      expect(roll.interpret({ raw: [5, 6, 10], difficulty: 6, explosions: [8] }).successes).toBe(3)
    })

    it('counts ones', () => {
      expect(roll.interpret({ raw: [5, 6, 10], explosions: [2] }).ones).toBe(0)
      expect(roll.interpret({ raw: [1, 2, 3] }).ones).toBe(1)
      expect(roll.interpret({ raw: [1, 5, 1], explosions: [1] }).ones).toBe(3)
    })

    it('counts maxes', () => {
      expect(roll.interpret({ raw: [5, 6, 7], sides: 10 }).maxes).toBe(0)
      expect(roll.interpret({ raw: [6, 4, 6], sides: 6 }).maxes).toBe(2)
      // sides defaults to d10s
      expect(roll.interpret({ raw: [10, 9, 4] }).maxes).toBe(1)
      expect(roll.interpret({ raw: [10, 9, 4], explosions: [10] }).maxes).toBe(2)
    })

    it('computes sums', () => {
      expect(roll.interpret({ raw: [5, 6, 7] }).sum).toBe(18)
      expect(roll.interpret({ raw: [1, 3, 5, 7] }).sum).toBe(16)
    })
  })

  describe('explode', () => {
    beforeEach(() => {
      roll.old_roll = roll.roll
      let times = 2
      roll.roll = jest.fn(() => (times-- >= 0 ? 10 : 5)) // eslint-disable-line no-plusplus
    })
    afterEach(() => {
      roll.roll = roll.old_roll
      delete roll.old_roll
    })

    it('returns [] if explode is not set', () => {
      expect(roll.explode([10, 10, 10], {})).toEqual([])
      expect(roll.roll).toHaveBeenCalledTimes(0)
    })
    it('explodes only once if explodeAt = !', () => {
      expect(roll.explode([10], { explode: '!', explodeAt: 10 })).toEqual([10])
    })
    it('explodes repeatedly if explodeAt = !!', () => {
      expect(roll.explode([10], { explode: '!!', explodeAt: 10 })).toEqual([10, 10, 10, 5])
    })
  })

  describe('handler', () => {
    const mockRolls = (rolls) => {
      let index = 0
      return jest.fn(() => rolls[index++]) // eslint-disable-line no-plusplus
    }

    beforeEach(() => {
      jest.clearAllMocks()

      roll.old_roll = roll.roll
      roll.old_interpret = roll.interpret

      roll.roll = mockRolls([5, 6, 1, 9, 10, 8, 7])
      roll.interpret = jest.fn((...args) => roll.old_interpret(...args))
    })
    afterEach(() => {
      roll.roll = roll.old_roll
      delete roll.old_roll

      roll.interpret = roll.old_interpret
      delete roll.old_interpret
    })

    it('parses count', () => {
      roll.handler('!roll 5')
      expect(roll.roll).toHaveBeenCalledTimes(5)

      jest.clearAllMocks()
      roll.handler('!roll 3d8')
      expect(roll.roll).toHaveBeenCalledTimes(3)
    })

    it('parses sides', () => {
      roll.handler('!roll 5d6')

      expect(roll.roll).toHaveBeenCalledTimes(5)
      roll.roll.mock.calls.forEach((_, index) => {
        expect(roll.roll).toHaveBeenNthCalledWith(index + 1, 6)
      })
    })

    it('defaults sides to 10', () => {
      roll.handler('!roll 3')

      expect(roll.roll).toHaveBeenCalledTimes(3)
      roll.roll.mock.calls.forEach((_, index) => {
        expect(roll.roll).toHaveBeenNthCalledWith(index + 1, 10)
      })
    })

    it('parses difficulty', () => {
      roll.handler('!roll 5@8')
      expect(roll.roll).toHaveBeenCalledTimes(5)
      expect(roll.interpret.mock.calls[0][0].difficulty).toEqual(8)

      jest.clearAllMocks()

      roll.handler('!roll 4d6@5')
      expect(roll.roll).toHaveBeenCalledTimes(4)
      expect(roll.interpret.mock.calls[0][0].sides).toEqual(6)
      expect(roll.interpret.mock.calls[0][0].difficulty).toEqual(5)
    })

    it('parses explode & explodeAt', () => {
      roll.roll = mockRolls([5, 6, 7, 8, 6])

      roll.handler('!roll 5!9')
      expect(roll.roll).toHaveBeenCalledTimes(5)
      expect(roll.interpret.mock.calls[0][0].explode).toEqual('!')
      expect(roll.interpret.mock.calls[0][0].explodeAt).toEqual(9)

      jest.clearAllMocks()

      roll.handler('!roll 2d10@8!!10')
      expect(roll.roll).toHaveBeenCalledTimes(2)
      expect(roll.interpret.mock.calls[0][0].count).toEqual(2)
      expect(roll.interpret.mock.calls[0][0].difficulty).toEqual(8)
      expect(roll.interpret.mock.calls[0][0].explode).toEqual('!!')
      expect(roll.interpret.mock.calls[0][0].explodeAt).toEqual(10)
      expect(roll.interpret.mock.calls[0][0].sides).toEqual(10)

      jest.clearAllMocks()

      roll.handler('!roll 5d8!!')
      expect(roll.roll).toHaveBeenCalledTimes(5)
      expect(roll.interpret.mock.calls[0][0].explode).toEqual('!!')
      expect(roll.interpret.mock.calls[0][0].explodeAt).toEqual(8)
      expect(roll.interpret.mock.calls[0][0].sides).toEqual(8)
    })
  })

  describe('formatMessage', () => {
    const mockRolls = (rolls) => {
      let index = 0
      return jest.fn(() => rolls[index++]) // eslint-disable-line no-plusplus
    }

    beforeEach(() => {
      jest.clearAllMocks()
      roll.old_roll = roll.roll
    })
    afterEach(() => {
      roll.roll = roll.old_roll
      delete roll.old_roll
    })

    it('displays correctly in each scenario', () => {
      roll.roll = mockRolls([1, 8, 5, 6])
      expect(roll.handler('!roll 4')).toMatchSnapshot()

      roll.roll = mockRolls([4, 8, 4, 5])
      expect(roll.handler('!roll 4d8')).toMatchSnapshot()

      roll.roll = mockRolls([9, 10, 10, 8, 6, 7, 5])
      expect(roll.handler('!roll 7d10')).toMatchSnapshot()

      roll.roll = mockRolls([1, 8, 5, 7, 9, 10, 8, 6, 4, 10])
      expect(roll.handler('!roll 9d10!')).toMatchSnapshot()
      expect(roll.roll).toHaveBeenCalledTimes(10)

      roll.roll = mockRolls([5, 10, 10, 6, 9, 10, 8, 8])
      expect(roll.handler('!roll 5d10!!')).toMatchSnapshot()
      expect(roll.roll).toHaveBeenCalledTimes(8)
    })
  })
})
