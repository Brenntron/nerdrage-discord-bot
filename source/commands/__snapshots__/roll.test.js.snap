// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`!roll formatMessage displays correctly in each scenario 1`] = `
"\`\`\`css
You rolled 4d10's -> ðŸŽ²[8, 6, 5, 1] + ðŸ’¥[none]
 -> [1] x #successes { @ 8 difficulty }
 -> [1] x #ones
 -> [0] x #10s
\`\`\`"
`;

exports[`!roll formatMessage displays correctly in each scenario 2`] = `
"\`\`\`css
You rolled 4d8's -> ðŸŽ²[8, 5, 4, 4] + ðŸ’¥[none]
 -> [1] x #successes { @ 6 difficulty }
 -> [0] x #ones
 -> [1] x #8s
\`\`\`"
`;

exports[`!roll formatMessage displays correctly in each scenario 3`] = `
"\`\`\`css
You rolled 7d10's -> ðŸŽ²[10, 10, 9, 8, 7, 6, 5] + ðŸ’¥[none]
 -> [4] x #successes { @ 8 difficulty }
 -> [0] x #ones
 -> [2] x #10s
\`\`\`"
`;

exports[`!roll formatMessage displays correctly in each scenario 4`] = `
"\`\`\`css
You rolled 9d10's -> ðŸŽ²[10, 9, 8, 8, 7, 6, 5, 4, 1] + ðŸ’¥[10]
 -> [5] x #successes { @ 8 difficulty }
 -> [1] x #ones
 -> [2] x #10s
\`\`\`"
`;

exports[`!roll formatMessage displays correctly in each scenario 5`] = `
"\`\`\`css
You rolled 5d10's -> ðŸŽ²[10, 10, 9, 6, 5] + ðŸ’¥[10, 8, 8]
 -> [6] x #successes { @ 8 difficulty }
 -> [0] x #ones
 -> [3] x #10s
\`\`\`"
`;
